flink-state-backends/flink-statebackend-forst/src/test/java/org/apache/flink/state/forst/ForStAsyncAndSyncCompatibilityTest.java

```diff
@@ -0,0 +1,234 @@
+        KeyedStateHandle stateHandle = snapshotResult.getJobManagerOwnedSnapshot();
+        IOUtils.closeQuietly(keyedBackend);
+        ForStSyncKeyedStateBackend<String> syncKeyedStateBackend =
+                createSyncKeyedStateBackend(
+                        forStStateBackend,
+                        env,
+                        StringSerializer.INSTANCE,
+                        Collections.singletonList(stateHandle));
+
+        try {
+            org.apache.flink.api.common.state.MapState<Integer, String> syncMapState =
+                    syncKeyedStateBackend.getOrCreateKeyedState(
+                            IntSerializer.INSTANCE,
+                            StateDescriptorUtils.transformFromV2ToV1(descriptor));
+            fail();
+
+            syncKeyedStateBackend.setCurrentKey("testKey");
+            ((InternalKvState) syncKeyedStateBackend).setCurrentNamespace(1);
+            Assert.assertEquals(syncMapState.get(1), "1");
+        } catch (Exception e) {
+            // Currently, ForStStateBackend does not support switching from Async to Sync, so this
+            // exception will be caught here
+            assertEquals(
+                    e.getMessage()
+                            .contains(
+                                    "org.apache.flink.runtime.state.v2.RegisteredKeyAndUserKeyValueStateBackendMetaInfo cannot be cast to class org.apache.flink.runtime.state.RegisteredKeyValueStateBackendMetaInfo"),
+                    true);
+        } finally {
+            IOUtils.closeQuietly(syncKeyedStateBackend);
+        }
+    }
+
+    @Test
+    void testForStTransFromSyncToAsync() throws Exception {
+        ForStSyncKeyedStateBackend<String> keyedBackend =
+                createSyncKeyedStateBackend(
+                        forStStateBackend, env, StringSerializer.INSTANCE, Collections.emptyList());
+        org.apache.flink.api.common.state.MapStateDescriptor<Integer, String> descriptor =
+                new org.apache.flink.api.common.state.MapStateDescriptor<>(
+                        "testState", IntSerializer.INSTANCE, StringSerializer.INSTANCE);
+        org.apache.flink.api.common.state.MapState<Integer, String> mapState =
+                keyedBackend.getOrCreateKeyedState(IntSerializer.INSTANCE, descriptor);
+        keyedBackend.setCurrentKey("testKey");
+        ((InternalKvState) mapState).setCurrentNamespace(1);
+        mapState.put(1, "1");
+
+        RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot =
+                keyedBackend.snapshot(
+                        1L,
+                        System.currentTimeMillis(),
+                        env.getCheckpointStorageAccess()
+                                .resolveCheckpointStorageLocation(
+                                        1L, CheckpointStorageLocationReference.getDefault()),
+                        CheckpointOptions.forCheckpointWithDefaultLocation());
+
+        if (!snapshot.isDone()) {
```

Comment:
I am curious what this line is doing? when will is be true and when will it be false?

Reply:
this line is wait for snapshot completing